### PR TITLE
Qtable discretize

### PR DESCRIPTION
**Pull request by chrisryda**
Implemented Discretizer as a class, with subclasses that allow for choosing which discretizer the snake should use

### CODE REVIEW

**Comment by chrisryda** [comment]
Gib review @tomrsae  @magsteen

File: discretizer.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    @property
+    def state_space_len(self) -> int:
+        wall_dist_x = self.clamped_wall_dists
+        wall_dist_y = self.clamped_wall_dists
+        tail_dist = self.clamped_tail_dists
+        n_directions = self.n_dirs
+        n_apple_dirs = self.n_apple_dirs
+        n_tail_dirs = self.n_tail_dirs
+        danger_flag = 2**4
+        print (wall_dist_x * wall_dist_y * n_directions * n_apple_dirs * n_tail_dirs * tail_dist * danger_flag)
```
Comment (tomrsae) [inline discretizer.py]: Remove debug print

File: snake_env.py [hunk 2]
```diff
@@ -46,25 +67,25 @@ def _turn_snake(self, action):
...
-            reward = 1 / self.steps
+            reward = 5 / self.steps
             self.steps = 0
         else:
             new_dist_to_apple = self.state.head_position.manhattan_dist(self.state.apple_position)
             reward = dist_to_apple - new_dist_to_apple
-            
         if won:
-            reward = 5
+            print(f"Homie just won... frfr. Time is {datetime.datetime.now()}")
```
Comment (tomrsae) [inline snake_env.py]: Remove debug print

File: snake_env.py [hunk 3]
```diff
@@ -46,25 +67,25 @@ def _turn_snake(self, action):
...
         won = self.state.has_won
-        reward = 0
-        self.steps += 1
         observation = self._get_obs()
         
+        self.steps += 1
+        reward = 0
         if ate:
-            reward = 1 / self.steps
+            reward = 5 / self.steps
```
Comment (tomrsae) [inline snake_env.py]: Should be `1 / self.steps`

File: snake_env.py [hunk 4]
```diff
@@ -22,6 +42,7 @@ def __init__(self, render_mode=None, seed=None, size=8):
         # Target's location, Neo's location and length
         self.observation_space = spaces.Dict({
```
Comment (tomrsae) [inline snake_env.py:43]: This needs to correspond with the changes made to obs space

**Review by tomrsae — CHANGES_REQUESTED** [review]
Looks great, just a few small things!

File: discretizer.py (hunk 1 above)
Comment (chrisryda) [reply discretizer.py]: Done!

File: snake_env.py (hunk 2 above)
Comment (chrisryda) [reply snake_env.py]: Done!

File: snake_env.py (hunk 3 above)
Comment (chrisryda) [reply snake_env.py]: Done!

File: snake_env.py (hunk 4 above)
Comment (chrisryda) [reply snake_env.py:43]: Done by Norbi!

Comment (tomrsae) [reply snake_env.py:43]: Not sufficient, still missing `grid`

Comment (chrisryda) [reply snake_env.py:43]: Done by Norbi!

**Review by tomrsae — APPROVED** [review]
:shipit:

**Pull request merged**